### PR TITLE
Fix #8317 - Warn in Output:Table:Monthly when invalid 'Variable or Meter Name' field

### LINKED ISSUE #8317: Output:Table:Monthly does not warn about invalid variable or meter name
Issue overview
--------------
Ran this poorly transitioned test file and saw now warning about this invalid meter name, just got zeros in the output table for "PEAK GAS DEMAND".
The offending meter name is "Water Heater:WaterSystems:Gas".

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 9.4-RC10

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here): https://github.com/NREL/EnergyPlusDevSupport/commit/1447601c30df8f07a5b4dd60864a5a36560c3fc9
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #8317
 - Warn in Output:Table:Monthly when invalid 'Variable or Meter Name' field

Using the defect file: https://github.com/NREL/EnergyPlusDevSupport/blob/1447601c30df8f07a5b4dd60864a5a36560c3fc9/DefectFiles/8000s/8317/1ZoneUncontrolled.idf#L453-L466

```
  Output:Table:Monthly,
    Building Loads - Cooling,!- Name
    2,                       !- Digits After Decimal
    Water Heater:WaterSystems:Gas,  !- Variable or Meter 1 Name       <============ This is wrong
    SumOrAverage,            !- Aggregation Type for Variable or Meter 1
    Zone Air System Sensible Cooling Rate,  !- Variable or Meter 2 Name
    Maximum,                 !- Aggregation Type for Variable or Meter 2
    Site Outdoor Air Drybulb Temperature,  !- Variable or Meter 3 Name
    ValueWhenMaximumOrMinimum;  !- Aggregation Type for Variable or Meter 3
```

```diff
(py38)julien@8317 (master *%>)$ diff -u out-ori/eplusout.err out-new/eplusout.err 

--- out-ori/eplusout.err	2020-10-19 11:30:00.923173743 +0200
+++ out-new/eplusout.err	2020-10-19 11:31:32.544227723 +0200
@@ -1,4 +1,4 @@
-Program Version,EnergyPlus, Version 9.4.0-998c4b761e, YMD=2020.10.19 11:30,
+Program Version,EnergyPlus, Version 9.4.0-6c1e5e5d73, YMD=2020.10.19 11:31,
    ** Warning ** Output:Meter: invalid Key Name="JUNKMETER:*" - not found.
    ** Warning ** Output:Meter: invalid Key Name="ELECTRICITY:JUNKRESOURCE" - not found.
    ************* Testing Individual Branch Integrity
@@ -9,6 +9,7 @@
    ************* All Return Air Paths passed integrity testing
    ************* No node connection errors were found.
    ************* Beginning Simulation
+   ** Warning ** In Output:Table:Monthly 'BUILDING LOADS - COOLING' invalid Variable or Meter Name 'WATER HEATER:WATERSYSTEMS:GAS'
    ************* Simulation Error Summary *************
    ** Warning ** The following Report Variables were requested but not generated -- check.rdd file
    **   ~~~   ** Either the IDF did not contain these elements, the variable name is misspelled,
@@ -16,4 +17,4 @@
    ************* Key=*, VarName=JUNK VARIABLE, Frequency=Monthly
    ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
    ************* EnergyPlus Sizing Error Summary. During Sizing: 0 Warning; 0 Severe Errors.
-   ************* EnergyPlus Completed Successfully-- 3 Warning; 0 Severe Errors; Elapsed Time=00hr 00min  0.51sec
+   ************* EnergyPlus Completed Successfully-- 4 Warning; 0 Severe Errors; Elapsed Time=00hr 00min  6.16sec
```

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

File: tst/EnergyPlus/unit/OutputReportTabular.unit.cc [hunk 1]
```diff
@@ -8339,3 +8339,96 @@ TEST_F(EnergyPlusFixture, OutputReportTabular_GatherHeatGainReport)
...
+        "  Exterior Lights Electric Power,   !- Variable or Meter 2 Name",   // A bad variable name
+        "  Maximum,                          !- Aggregation Type for Variable or Meter 2",
+        "  AlwaysOn,                         !- Variable or Meter 3 Name",   // A bad name (eg: Schedule)
+        "  Maximum,                          !- Aggregation Type for Variable or Meter 3",
+        "  Exterior Lights Electricity Rate, !- Variable or Meter 4 Name",   // A good variable name
+        "  Minimum,                          !- Aggregation Type for Variable or Meter 4",
+        "  OnSched,                          !- Variable or Meter 5 Name",   // A good schedule name
+        "  Minimum,                          !- Aggregation Type for Variable or Meter 5",
+        "  Heating:NaturalGas,               !- Variable or Meter 6 Name",   // A good meter name
+        "  SumOrAverage;                     !- Aggregation Type for Variable or Meter 6",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/OutputReportTabular.unit.cc:8376]: Unit test, testing several possibilities

File: tst/EnergyPlus/unit/OutputReportTabular.unit.cc [hunk 2]
```diff
@@ -8339,3 +8339,96 @@ TEST_F(EnergyPlusFixture, OutputReportTabular_GatherHeatGainReport)
...
+    OutputReportTabular::GetInputTabularMonthly(state);
+    OutputReportTabular::InitializeTabularMonthly(state);
+
+    std::string expected_error = delimited_string({
+        "   ** Warning ** In Output:Table:Monthly 'MY REPORT' invalid Variable or Meter Name 'HEATING:GAS'",
+        "   ** Warning ** In Output:Table:Monthly 'MY REPORT' invalid Variable or Meter Name 'EXTERIOR LIGHTS ELECTRIC POWER'",
+        "   ** Warning ** In Output:Table:Monthly 'MY REPORT' invalid Variable or Meter Name 'ALWAYSON'",
+    });
+
+    compare_err_stream(expected_error);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/OutputReportTabular.unit.cc:8433]: Check that you do get the warnings as expected

File: src/EnergyPlus/OutputProcessor.cc [hunk 3]
```diff
@@ -6823,6 +6819,29 @@ void UpdateMeterReporting(EnergyPlusData &state)
...
+                {
+                    meterIndex = Meter;
+                    break;
+                }
+            }
+        }
+
+        return meterIndex;
+    };
+
```
Comment (jmarrec) [inline src/EnergyPlus/OutputProcessor.cc:6841]: I initially went for a solution were I was reusing this routine (among with checking if it was an output:variable or a Schedule) so I made it its own function and dried up the code below as well. I'm not using it, and it's not going to be used anywhere else, so I made it a lambda.

File: src/EnergyPlus/OutputProcessor.cc [hunk 4]
```diff
@@ -6844,33 +6863,14 @@ void UpdateMeterReporting(EnergyPlusData &state)
...
-            }
-            if (NeverFound) {
-                ShowWarningError(cCurrentModuleObject + ": invalid " + cAlphaFieldNames(1) + "=\"" + Alphas(1) + "\" - not found.");
-            }
+        int meterIndex = findMeterIndexFromMeterName(Alphas(1));
+        if (meterIndex > 0) {
+            // MeterFileOnlyIndicator is false, CumulativeIndicator is false
+            SetInitialMeterReportingAndOutputNames(state, meterIndex, false, ReportFreq, false);
+        } else {
+            ShowWarningError(cCurrentModuleObject + ": invalid " + cAlphaFieldNames(1) + "=\"" + Alphas(1) + "\" - not found.");
```
Comment (jmarrec) [inline src/EnergyPlus/OutputProcessor.cc:6870]: The use of the (lambda) function allows to dry up this code block and 3 others below (4 total)

File: src/EnergyPlus/OutputReportTabular.cc [hunk 5]
```diff
@@ -1148,6 +1148,9 @@ namespace OutputReportTabular {
                 curVariMeter = UtilityRoutines::MakeUPPERCase(MonthlyFieldSetInput(FirstColumn + colNum - 1).variMeter);
                 // call the key count function but only need count during this pass
                 GetVariableKeyCountandType(state, curVariMeter, KeyCount, TypeVar, AvgSumVar, StepTypeVar, UnitsVar);
+                if (TypeVar == OutputProcessor::VarType_NotFound) {
+                    ShowWarningError("In Output:Table:Monthly '" +  MonthlyInput(TabNum).name + "' invalid Variable or Meter Name '" + curVariMeter + "'");
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/OutputReportTabular.cc:1150]: This is the actual warning. After going the manual route, I noticed that `GetVariableKeyCountandType` was already returning all the information needed to be able to throw a warning, so might as well use it here.

**Review by mjwitte — APPROVED** [review]
@jmarrec This looks great. I've merged in develop for one last round of CI tests, then this will merge.

**Pull request merged**